### PR TITLE
Fix failing builds to bump Gunicorn, Django, and gevent versions

### LINKED ISSUE #51: Upgrade to Django 2.0.6
See https://www.djangoproject.com/weblog/2018/jun/01/bugfix-release/

### PR DESCRIPTION
**Pull request by rbreslow**
## Overview

Reviewing #52 revealed issues that I feel are out of the scope of that PR.

Builds for slim images were failing:

```bash
Step 3/9 : RUN apt-key adv --keyserver ha.pool.sks-keyservers.net --recv-keys B97B0AFCAA1A47F044F244A07FCC7D46ACCC4CF8
 ---> Running in 607bb7823ea0
E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation
The command '/bin/sh -c apt-key adv --keyserver ha.pool.sks-keyservers.net --recv-keys B97B0AFCAA1A47F044F244A07FCC7D46ACCC4CF8' returned a non-zero code: 255
```

This was unsettling because the previous build was [successful](https://travis-ci.org/azavea/docker-django/builds/373973299) and the only change I made was bumping the Django version in `2.0/requirements.txt`. I didn't make any change that would have removed `gpg`.

tianon/docker-brew-debian#49 was opened by someone who appears to be a Debian maintainer:

> Due to a change in APT (such that it no longer Depends: gnupg given that gnupg is only required for the optional apt-key functionality, and is thus now relegated to a Recommends relationship which deboostrap will not install), gnupg is no longer included in our sid/stretch images. Reading through #830696, this is firmly by-design (since otherwise, building a chroot without gpg would be a challenge).

The issue references an exchange from the Debian mailing list:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=830696

A commit from 2016 changes the `apt` package relationship for `gnupg` from `Depends` to `Recommends`:

- https://salsa.debian.org/apt-team/apt/blob/master/debian/changelog#L1102
- https://salsa.debian.org/apt-team/apt/commit/87d468fe355c87325c943c40043a0bb236b2407f

This change was introduced in `apt (1.3~exp1) experimental`.

13 days ago, https://github.com/docker-library/official-images/commit/f200d7ffdeb698ad62802351dd9e023cad57a4bd removed `3.5-slim` and updated `2.7-slim` and `3.6-slim` to derive from `debian:stretch-slim` instead of `debian:jessie-slim`.

Jessie [is built with](https://github.com/debuerreotype/docker-debian-artifacts/blob/b024a792c752a5c6ccc422152ab0fd7197ae8860/jessie/slim/rootfs.manifest#L3) `apt 1.0.9.8.4`, which was introduced before the `gnupg` package relationship was broken:

```bash
> rbreslow@maiden docker-django (feature/jrb/django-updates) $ docker run --entrypoint apt d6f -v

WARNING: apt does not have a stable CLI interface yet. Use with caution in scripts.

apt 1.0.9.8.4 for amd64 compiled on Dec 11 2016 09:48:19
...
```

While Stretch [is built with](https://github.com/debuerreotype/docker-debian-artifacts/blob/b024a792c752a5c6ccc422152ab0fd7197ae8860/stretch/slim/rootfs.manifest#L2) `apt 1.4.8`:

```bash
> rbreslow@maiden docker-django (feature/jrb/django-updates) $ docker run --entrypoint apt f5d -v
apt 1.4.8 (amd64)
```

This explains why the slim-variant builds began to fail.

I borrowed a [fix](https://github.com/docker-library/postgres/blob/3f8d33c06e25295260a6d30defa6dce40aea8ad8/10/Dockerfile#L4) from [docker-library/postgres](https://github.com/docker-library/postgres), and refactored our `Dockerfile` templates to incorporate the new changes to Python images, as well as what I identified as best practice.

This was helpful reference:
 - https://github.com/docker-library/gcc/commit/00013aa569bfe01e32300b8eb70db417a43a7bb5
 - https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache

Alpine images were also failing due to an update to `gdal` that introduced a missing dependency. I updated `Dockerfile-alpine.template` to install `libressl2.7-libcrypto`. 

See:
 - appropriate/docker-postgis#96
 - appropriate/docker-postgis#97

Finally, I bumped the Gunicorn, Django, and gevent versions.

See:
 - http://docs.gunicorn.org/en/stable/news.html
 - https://www.djangoproject.com/weblog/2018/jun/01/bugfix-release/
 - http://www.gevent.org/changelog.html

Fixes #51

## Notes
Alpine Linux removed all PostgresSQL 9.x packages from their `3.7` and `edge` apk repositories.

Instead of pulling `postgresql-client` on the slim variant, I pull `postgresql-client-$PG_MAJOR`. This is what I interpreted as best practice described by https://wiki.postgresql.org/wiki/Apt, and allowed me to source PostgreSQL 9.4 from the official apt. 

I updated the README to reflect these changes, as well as to reflect changes to the Python images on Docker Hub (removal of `3.5-slim`). 

I am not sure if this is what you would consider best practice, @hectcastro, and I am aware that this changes core functionality. I am not sure if we have any services dependent on Python 3.5, etc. I would appreciate your feedback on my approach to solving the issues I describe in my PR.

## Testing

See Travis CI build: https://travis-ci.org/azavea/docker-django/builds/388480749

### CODE REVIEW

File: 1.10/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-gunicorn==19.7.1
+gunicorn==19.8.1
 django==1.10.8
 psycopg2==2.7.4 --no-binary psycopg2
-gevent==1.2.2
+gevent==1.3.2.post0
```
Comment (hectcastro) [inline 1.10/requirements.txt:4]: Just to confirm, you're using the `1.3.2.post0` package because the `1.3.2` one was busted?

**Comment by rbreslow** [comment]
Some of the Travis CI builds failed with an error similar to nodejs/docker-node#380. The jobs seem to have restarted themselves and recovered. Not sure what's going on?

File: README.md [hunk 2]
```diff
@@ -6,17 +6,17 @@ This repository contains a collection of templated `Dockerfile` for image varian
 
 ### Template Variables
 
-- `PYTHON_VERSION` - Python interpreter version
-- `PG_MAJOR` - Major version number of the target PostgreSQL database instance
-- `PG_VERSION` - Version number for `postgresql-client` package
-- `VARIANT` - Base container image variant
+- `PYTHON_VERSION` - Python interpreter version (for `python:*-slim-stretch` or `python:*-alpine` [images](https://hub.docker.com/_/python/))
```
Comment (hectcastro) [inline README.md:9]: Good `README` additions.

File: Dockerfile-slim.template [hunk 3]
```diff
@@ -19,11 +38,12 @@ RUN set -ex \
...
        gettext \
-       postgresql-client=${PG_VERSION} \
+       postgresql-client-$PG_MAJOR=$PG_VERSION \
     " \
-    && apt-get update && apt-get install -y ${buildDeps} ${deps} --no-install-recommends \
+    && apt-get update && apt-get install -y $buildDeps $deps --no-install-recommends \
     && pip install --no-cache-dir -r /tmp/requirements.txt \
-    && apt-get purge -y --auto-remove ${buildDeps} \
+    && apt-get purge -y --auto-remove $buildDeps \
+       $(! command -v gpg > /dev/null || echo 'gnupg dirmngr') \
```
Comment (hectcastro) [inline Dockerfile-slim.template:46]: What is the purpose of this?

File: README.md [hunk 4]
```diff
@@ -6,17 +6,17 @@ This repository contains a collection of templated `Dockerfile` for image varian
...
+- `VARIANT` - Base container image variant (`alpine` or `slim`)
 
 ### Testing
 
 An example of how to use `cibuild` to build and test an image:
 
 ```bash
-$ CI=1 VERSION=1.9 PYTHON_VERSION=2.7 \
-  PG_MAJOR=9.6 PG_VERSION=9.5.4-r0 VARIANT=alpine \
+$ CI=true VERSION=2.0 PYTHON_VERSION=3.6 \
```
Comment (hectcastro) [inline README.md]: Although both work based on how we check, the use of integers as boolean flags seems more of a convention in Bash scripting than the strings `true` or `false`.

File: Dockerfile-slim.template [hunk 5]
```diff
@@ -1,14 +1,33 @@
-FROM python:%%PYTHON_VERSION%%-slim
+FROM python:%%PYTHON_VERSION%%-slim-stretch
```
Comment (hectcastro) [inline Dockerfile-slim.template:1]: Making this more explicit is a good idea.

**Comment by rbreslow** [comment]
Saw what might be this in the CI output:

moby/moby#29005

File: Dockerfile-alpine.template [hunk 6]
```diff
@@ -10,11 +10,13 @@ COPY requirements.txt /tmp/
...
        gcc \
-       postgresql-dev=${PG_VERSION} \
+       postgresql-dev=$PG_VERSION \
        musl-dev \
     && apk add --no-cache \
        gettext \
-       postgresql-client=${PG_VERSION} \
+       postgresql-client=$PG_VERSION \
+    && apk add --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/main \
+        libressl2.7-libcrypto \
```
Comment (hectcastro) [inline Dockerfile-alpine.template:19]: This is confusing, but appears to be a required upgrade from 2.6 given that `gdal` is coming from `edge`. I wasn't able to make it work with another combination. 😞

File: Dockerfile-slim.template [hunk 7]
```diff
@@ -1,14 +1,33 @@
...
+		apt-get update; \
+		apt-get install -y --no-install-recommends \
+			gnupg \
+			dirmngr \
+		; \
+		rm -rf /var/lib/apt/lists/*; \
+	fi
+
+RUN set -ex; \
+# pub   4096R/ACCC4CF8 2011-10-13 [expires: 2019-07-02]
```
Comment (rbreslow) [inline Dockerfile-slim.template:16]: I think this might be causing moby/moby#29005

File: .travis.yml [hunk 8]
```diff
@@ -2,20 +2,20 @@ language: bash
...
-  - VERSION=1.10 PYTHON_VERSION=2.7 PG_MAJOR=9.6 PG_VERSION=9.5.12-r0 VARIANT=alpine
-  - VERSION=1.10 PYTHON_VERSION=3.5 PG_MAJOR=9.6 PG_VERSION=9.5.12-r0 VARIANT=alpine
-  - VERSION=1.11 PYTHON_VERSION=2.7 PG_MAJOR=9.6 PG_VERSION=10+190.pgdg80+1 VARIANT=slim
-  - VERSION=1.11 PYTHON_VERSION=3.6 PG_MAJOR=9.6 PG_VERSION=10+190.pgdg80+1 VARIANT=slim
-  - VERSION=1.11 PYTHON_VERSION=2.7 PG_MAJOR=9.6 PG_VERSION=9.5.12-r0 VARIANT=alpine
-  - VERSION=1.11 PYTHON_VERSION=3.6 PG_MAJOR=9.6 PG_VERSION=9.5.12-r0 VARIANT=alpine
-  - VERSION=2.0 PYTHON_VERSION=3.6 PG_MAJOR=9.6 PG_VERSION=10+190.pgdg80+1 VARIANT=slim
-  - VERSION=2.0 PYTHON_VERSION=3.6 PG_MAJOR=9.6 PG_VERSION=9.5.12-r0 VARIANT=alpine
+  - VERSION=1.9 PYTHON_VERSION=2.7 PG_MAJOR=9.4 PG_VERSION=9.4.18-2.pgdg90+1 VARIANT=slim
+  - VERSION=1.9 PYTHON_VERSION=3.6 PG_MAJOR=9.4 PG_VERSION=9.4.18-2.pgdg90+1 VARIANT=slim
```
Comment (hectcastro) [inline .travis.yml]: 1.9 does not support Python 3.6, so I think that we should drop those items from the matrix (the image will still exist in Quay, it just won't get updated).

See: https://docs.djangoproject.com/en/2.0/faq/install/#what-python-version-can-i-use-with-django

File: .travis.yml [hunk 9]
```diff
@@ -2,20 +2,20 @@ language: bash
...
+  - VERSION=1.9 PYTHON_VERSION=2.7 PG_MAJOR=10 PG_VERSION=10.4-r0 VARIANT=alpine
+  - VERSION=1.9 PYTHON_VERSION=3.6 PG_MAJOR=10 PG_VERSION=10.4-r0 VARIANT=alpine
+  - VERSION=1.10 PYTHON_VERSION=2.7 PG_MAJOR=10 PG_VERSION=10.4-2.pgdg90+1 VARIANT=slim
+  - VERSION=1.10 PYTHON_VERSION=3.6 PG_MAJOR=10 PG_VERSION=10.4-2.pgdg90+1 VARIANT=slim
```
Comment (hectcastro) [inline .travis.yml]: Same here. 1.10 does not support Python 3.6.

**Review by hectcastro — COMMENTED** [review]
What a great time to inherit the maintenance of this container image! Good work getting to the root of the `jessie` -> `stretch` swap and making the appropriate fixes.

Just to surface an alternate approach for discussion, would it have been possible to just pin things to `jessie` for now and deal with all of the changes required to support `stretch` later?

**Comment by hectcastro** [comment]
> The jobs seem to have restarted themselves and recovered. Not sure what's going on?

I restarted the 2 builds that were failing manually via the Travis UI.

File: 1.10/requirements.txt (hunk 1 above)
Comment (rbreslow) [reply 1.10/requirements.txt:4]: Yeah. There was a [packaging error](http://www.gevent.org/changelog.html#post0-2018-05-30) I didn't dive too deep into.

File: Dockerfile-slim.template (hunk 3 above)
Comment (rbreslow) [reply Dockerfile-slim.template:46]: Removes `gnupg` and `dirmngr` if `gpg` exists (exits zero).

File: README.md (hunk 4 above)
Comment (rbreslow) [reply README.md]: 👍

File: Dockerfile-alpine.template (hunk 6 above)
Comment (rbreslow) [reply Dockerfile-alpine.template:19]: Yeah, this was my experience as well. 

Want to resurface these:

 - appropriate/docker-postgis#96
 - appropriate/docker-postgis#97

File: Dockerfile-slim.template (hunk 5 above)
Comment (rbreslow) [reply Dockerfile-slim.template]: 👍

File: .travis.yml (hunk 8 above)
Comment (rbreslow) [reply .travis.yml]: 👍

File: .travis.yml (hunk 9 above)
Comment (rbreslow) [reply .travis.yml]: 👍

**Comment by rbreslow** [comment]
>I restarted the 2 builds that were failing manually via the Travis UI.

Got it. 

>Just to surface an alternate approach for discussion, would it have been possible to just pin things to `jessie` for now and deal with all of the changes required to support `stretch` later?

I didn't connect the issues with slim builds to a change in Debian distribution until looking at the history of the `apt` repo, fixing `gnupg`, and then tying it together with [debuerreotype/docker-debian-artifacts](https://github.com/debuerreotype/docker-debian-artifacts/tree/b024a792c752a5c6ccc422152ab0fd7197ae8860) and https://github.com/docker-library/official-images/commit/f200d7ffdeb698ad62802351dd9e023cad57a4bd. 

In retrospect, knowing that `*-slim` had it's distribution swapped would have allowed me to complete #52 and open an issue to upgrade this repository to `stretch`.

**Comment by rbreslow** [comment]
@hectcastro Let me know if https://github.com/azavea/docker-django/pull/53/commits/2a2c54c3dac150cbaf4a9aaa53e77d305974e0bf resolves compatibility.

File: Dockerfile-slim.template (hunk 3 above)
Comment (hectcastro) [reply Dockerfile-slim.template:46]: Ah, I see now. Thanks for clarifying.

**Review by hectcastro — APPROVED** [review]
👍

**Pull request merged**